candies/cli/cli.py

```diff
@@ -57,6 +61,60 @@ def help(self, *args, **kwargs):
         """
         return self.main.help(*args, **kwargs)
 
+    def parse(self, callable: Union[Callable[[str], Any], Type]) \
+            -> Union[Type, Callable[[str], Any]]:
+        """Defines a parser for a custom type.
+
+        Args:
+            callable: A callable that converts a string to an instance
+                of a custom type. For example, a parser function or a
+                type that accepts a string to a constructor.
+
+        Examples:
+            @cli
+            def add(money: Arg[Money]):
+                ...
+
+            @main.parse
+            def money(x: str) -> Money:
+                if x[0] == '$':
+                    return Money(float(x[1:]), 'dollars')
+                ...
+
+            # Or, as an alternative, if
+            # `Money.__init__(str)` can be called.
+            main.parse(Money)
+        """
+
+        if isinstance(callable, type):
+            type_ = callable
+        else:
+            signature = inspect.signature(callable)
+
+            if len(signature.parameters) != 1:
```

Comment:
Should we check the annotation of the single parameter?

To address something like that:
```py
@main.parse
def money(x: type) -> Money:
    ...
```

Reply:
At first I thought it would be nice to add such a check (and I actually added it), but then I thought that it doesn't really matter what the user annotated this parameter with. We will pass a string anyways, so we have no reason to add any restrictions.

Reply:
Yeah, I was thinking in this way too. However, the key point for me is that the wrong type annotation may lead to wrong linting and autocompletion as well as indirect error message.

For example:
```py
class Thing:
    def __init__(self, name: str):
        self.name = name

@main.parse
def money(x: Thing) -> Money:
    print(x.name)

    return Money.dollars(0)
```

I'm just wondering, whether one level of indirection is okay, comparing messages `'str' object has no attribute 'name'` and `Parsing function 'money' has invalid signature: it's expecting 'Thing' as the first argument instead of 'str'`.

Not sure, maybe this is just an overkill. I just like as helpful errors as possible. ))

Reply:
Agree. Let's allow the annotation to be either empty or `str`. Otherwise, throw an exception. What do you think?